sdc/rewrites/dataframe_getitem_attribute.py

```diff
@@ -0,0 +1,87 @@
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *****************************************************************************
+
+from numba.ir import Assign, Const, Expr, Var
+from numba.ir_utils import mk_unique_var
+from numba.rewrites import register_rewrite, Rewrite
+from numba.types import StringLiteral
+from numba.typing import signature
+
+from sdc.config import config_pipeline_hpat_default
+from sdc.hiframes.pd_dataframe_type import DataFrameType
+
+
+if not config_pipeline_hpat_default:
+    @register_rewrite('after-inference')
+    class RewriteDataFrameGetItemAttr(Rewrite):
+        """
+        Search for calls of df.attr and replace it with calls of df['attr']:
+        $0.2 = getattr(value=df, attr=A) -> $const0.0 = const(str, A)
+                                            $0.2 = static_getitem(value=df, index=A, index_var=$const0.0)
+        """
+
+        def match(self, func_ir, block, typemap, calltypes):
+            self.block = block
+            self.typemap = typemap
+            self.calltypes = calltypes
+            self.getattrs = getattrs = set()
+            for expr in block.find_exprs(op='getattr'):
+                obj = typemap[expr.value.name]
+                if not isinstance(obj, DataFrameType):
+                    continue
+                if expr.attr in obj.columns:
+                    getattrs.add(expr)
+
+            return len(getattrs) > 0
+
+        def apply(self):
+            new_block = self.block.copy()
+            new_block.clear()
+            for inst in self.block.body:
+                if isinstance(inst, Assign):
+                    expr = inst.value
+                    if expr in self.getattrs:
+                        const_assign = self._assign_const(inst)
+                        new_block.append(const_assign)
+
+                        new_expr = Expr.getitem(expr.value, const_assign.target, inst.loc)
+                        self.calltypes[new_expr] = signature(
+                            self.typemap[inst.target.name],
+                            self.typemap[expr.value.name],
+                            self.typemap[new_expr.index.name]
+                        )
+                        inst = Assign(value=new_expr, target=inst.target, loc=inst.loc)
+                new_block.append(inst)
+
+            return new_block
+
+        def _assign_const(self, inst, prefix='$const0'):
+            """Create constant from attribute of specified instruction."""
+            const_node = Const(inst.value.attr, inst.loc)
+            const_var = Var(inst.target.scope, mk_unique_var(prefix), inst.loc)
```

Comment:
You could use `declare_constant` from ir_utils:

```python
declare_constant(inst.value.attr, self.block, self.func_ir, inst.loc)
```

This function also adds this constant to the beginning of the block. So, you don't need to do `new_block.append(const_assign)`

Reply:
When you are declaring variable (including variables assigned to constants) you must update `func_ir._definitions` dictionary. Otherwise some functionality wouldn't work and this may result in errors lately. (btw `declare_constant` updates this dictionary)

Reply:
```python
func_ir._definitions[const_var.name] = [const_node]
```